src/transformation/mesh_intersection/mesh_intersection.rs

```diff
@@ -256,380 +355,358 @@ fn extract_connected_components(
-                                .delaunay
-                                .add_constraint(handles_a[i], handles_b[i]);
-                        }
-                    }
-                }
-                TriangleTriangleIntersection::Polygon(intersections) => {
-                    for inter in intersections {
-                        let f1 = convert_fid(mesh1, i1, inter.f1);
-                        let f2 = convert_fid(mesh2, i2, inter.f2);
-                        let orig_fid = [inter.f1, inter.f2];
-                        let key = (f1, f2);
-                        let ins = *intersection_points
-                            .entry(key)
-                            .or_insert([inter.p1, inter.p2]);
-
-                        let _ = insert_point(ins, key, orig_fid, 0);
-                        let _ = insert_point(ins, key, orig_fid, 1);
-                    }
-                }
+fn insert_into_set(
+    position: Vector3<f64>,
+    point_set: &mut RTree<TreePoint>,
+    epsilon: f64,
+) -> usize {
+    let point_count = point_set.size();
+    let point_to_insert = TreePoint {
+        point: position,
+        id: point_count,
+    };
+
+    match point_set.nearest_neighbor(&point_to_insert) {
+        Some(tree_point) => {
+            if (tree_point.point - position).norm_squared() <= epsilon {
+                return tree_point.id;
+            } else {
+                point_set.insert(point_to_insert);
+                debug_assert!(point_set.size() == point_count + 1);
+                return point_count;
             }
         }
+        None => {
+            point_set.insert(point_to_insert);
+            debug_assert!(point_set.size() == point_count + 1);
+            return point_count;
+        }
     }
-
-    extract_result(
-        pos12,
-        mesh1,
-        flip1,
-        mesh2,
-        flip2,
-        &spade_handle_to_intersection,
-        &intersection_points,
-        &triangulations1,
-        &triangulations2,
-        new_vertices12,
-        new_indices12,
-    );
 }
 
-fn convert_fid(mesh: &TriMesh, tri: u32, fid: FeatureId) -> FeatureId {
-    match fid {
-        FeatureId::Edge(eid) => {
-            let topology = mesh.topology().unwrap();
-            let half_edge_id = topology.face_half_edges_ids(tri)[eid as usize];
-            let half_edge = &topology.half_edges[half_edge_id as usize];
-            // NOTE: if the twin doesn’t exist, it’s equal to u32::MAX. So the `min` will
-            //       automatically filter it out.
-            FeatureId::Edge(half_edge_id.min(half_edge.twin))
+fn smallest_angle(points: &[Vector3<f64>]) -> f64 {
+    let n = points.len();
+
+    let mut worst_cos = 2.0;
+    for i in 0..points.len() {
+        let d1 = (points[i] - points[(i + 1) % n]).normalize();
+        let d2 = (points[(i + 2) % n] - points[(i + 1) % n]).normalize();
+
+        let cos = d1.dot(&d2);
+
+        if cos < worst_cos {
+            worst_cos = cos.abs();
```

Comment:
Is it normal that `cos < worst_cos` doesn’t take the absolute value of `cos`?

Reply:
You are right that there is something wonky here, but it is not the absolute value. We are looking for the smallest angle, so the largest cos. The actual logic should have been grabbing the maximum of all unsigned cos. Thank you for catching this, I was quite lucky the algorithm was working despite the bug.

```rs
fn smallest_angle(points: &[Point3<Real>]) -> Real {
    let n = points.len();

    let mut worst_cos = -2.0;
    for i in 0..points.len() {
        let d1 = (points[i].coords - points[(i + 1) % n].coords).normalize();
        let d2 = (points[(i + 2) % n].coords - points[(i + 1) % n].coords).normalize();

        let cos = d1.dot(&d2);

        if cos > worst_cos {
            worst_cos = cos;
        }
    }

    worst_cos.acos()
}
```

Reply:
The problem is actually that I should have been looking for the largest signed `cos`. Very lucky the code works despite this bug.

```rs
fn smallest_angle(points: &[Point3<Real>]) -> Real {
    let n = points.len();

    let mut worst_cos = -2.0;
    for i in 0..points.len() {
        let d1 = (points[i].coords - points[(i + 1) % n].coords).normalize();
        let d2 = (points[(i + 2) % n].coords - points[(i + 1) % n].coords).normalize();

        let cos = d1.dot(&d2);
        if cos > worst_cos {
            worst_cos = cos;
        }
    }

    worst_cos.acos()
}
```